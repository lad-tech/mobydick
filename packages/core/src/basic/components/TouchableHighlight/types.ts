import {
  TouchableHighlightProps as DefaultTouchableHighlightProps,
  View,
} from 'react-native';

export type TouchableHighlightProps = DefaultTouchableHighlightProps;
export type ITouchableHighlight = View & TouchableHighlightProps;
