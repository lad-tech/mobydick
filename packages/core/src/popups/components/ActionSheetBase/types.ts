import {StyleProp, ViewStyle} from 'react-native';
import {PropsWithChildren, ReactElement} from 'react';

import {TypographyProp} from '../../../typography/types';
import {IPopup} from '../../types';

export interface IPropsContents {
  title: string;
  textFont?: TypographyProp;
  leftIcon?: ReactElement | undefined;
  onPress?(): void;
  radio?: string;
  checkboxList?: string[];
  disabled?: boolean;
  itemType: IItemType;
}

export enum IItemType {
  firstItem = 'firstItem',
  innerItem = 'innerItem',
  lastItem = 'lastItem',
  singleItem = 'singleItem',
  cancelItem = 'cancelItem',
}

export interface IPropsItem extends IPropsContents {
  style?: ViewStyle;
}

export interface IActionSheetBaseProps
  extends PropsWithChildren<Omit<IPopup, 'Content'>> {
  onClose: () => void;
  overlayStyle?: StyleProp<ViewStyle>;
  containerStyle?: StyleProp<ViewStyle>;
}
