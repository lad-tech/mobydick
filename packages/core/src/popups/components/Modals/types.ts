import {StyleProp, ViewStyle} from 'react-native';

import {IButtonTypes} from '../../../cta';

export interface IModalProps {
  title: string;
  descriptionText: string;
  buttonText?: string;
}

export interface IModalAskProps {
  title: string;
  descriptionText: string;

  typeLeft?: IButtonTypes;
  textLeft?: string;

  onPressRight(): void;
  typeRight?: IButtonTypes;
  textRight?: string;

  overlayStyle?: StyleProp<ViewStyle>;
  containerStyle?: StyleProp<ViewStyle>;
}
