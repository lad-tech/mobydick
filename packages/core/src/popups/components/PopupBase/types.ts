import {StyleProp, TextStyle, ViewStyle} from 'react-native';

import {TypographyProp} from '../../../typography/types';

export interface ITitlePopup {
  title: string;
  titleStyles?: TextStyle | TextStyle[];
  titleFont?: TypographyProp;
}

export interface IDescriptionTextPopup {
  descriptionText: string;
  descriptionStyles?: TextStyle | TextStyle[];
  descriptionFont?: TypographyProp;
}

export interface ITextContentProps {
  title?: string | undefined;
  titleStyles?: TextStyle | TextStyle[];
  titleFont?: TypographyProp;

  descriptionText?: string | undefined;
  descriptionStyles?: TextStyle | TextStyle[];
  descriptionFont?: TypographyProp;
}

export interface IPopupCloseIcon {
  onPress(): void;
}

export type IPopupProps = {
  overlayStyle?: StyleProp<ViewStyle>;
  onClose: () => void;
};
