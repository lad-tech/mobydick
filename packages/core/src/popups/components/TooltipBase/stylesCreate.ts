import {IPosition} from '../../types';
import {createStyles} from '../../../styles';

const stylesCreate = createStyles(({spaces, colors}, position?: IPosition) => {
  const positionStyle = () => {
    switch (position) {
      case IPosition.top:
        return {
          top: -spaces.Space8,
          borderBottomWidth: spaces.Space8,
          borderRightWidth: spaces.Space8,
          borderTopWidth: 0,
          borderLeftWidth: spaces.Space8,
          borderBottomColor: colors.BgInverse,
          borderRightColor: 'transparent',
          borderTopColor: 'transparent',
          borderLeftColor: 'transparent',
        };
      case IPosition.bottom:
      default:
        return {
          bottom: -spaces.Space8,
          borderTopWidth: spaces.Space8,
          borderRightWidth: spaces.Space8,
          borderBottomWidth: 0,
          borderLeftWidth: spaces.Space8,
          borderTopColor: colors.BgInverse,
          borderRightColor: 'transparent',
          borderBottomColor: 'transparent',
          borderLeftColor: 'transparent',
        };
    }
  };
  return {
    container: {
      backgroundColor: colors.BgInverse,
      paddingVertical: spaces.Space8,
      paddingHorizontal: spaces.Space16,
      borderRadius: spaces.Space8,
      justifyContent: 'center',
      alignItems: 'flex-start',
      zIndex: 1,
      marginVertical: spaces.Space8,
      position: 'absolute',
    },
    overlayStyle: {
      backgroundColor: 'transparent',
    },
    arrow: {
      position: 'absolute',
      width: 0,
      height: 0,
      backgroundColor: 'transparent',
      borderStyle: 'solid',
    },
    title: {
      zIndex: 1,
    },
    descriptionText: {
      zIndex: 1,
      paddingBottom: spaces.Space8,
    },
    positionStyle: {
      ...positionStyle(),
    },
  };
});

export default stylesCreate;
