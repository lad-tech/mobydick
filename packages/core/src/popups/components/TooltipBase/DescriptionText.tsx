import {FC} from 'react';

import {IDescriptionTextPopup} from '../PopupBase';
import useStyles from '../../../styles/hooks/useStyles';
import {Typography} from '../../../typography';

import stylesCreate from './stylesCreate';

const DescriptionText: FC<IDescriptionTextPopup> = props => {
  const [styles] = useStyles(stylesCreate);
  const {descriptionText, descriptionStyles, descriptionFont} = props;

  return (
    <Typography
      font={descriptionFont ? descriptionFont : 'Regular-Inverse-S'}
      style={[styles.descriptionText, descriptionStyles]}>
      {descriptionText}
    </Typography>
  );
};

export default DescriptionText;
