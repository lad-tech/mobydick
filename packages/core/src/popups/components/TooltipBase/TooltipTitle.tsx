import {FC} from 'react';

import {ITitlePopup} from '../PopupBase';
import useStyles from '../../../styles/hooks/useStyles';
import {TypographyLegacy} from '../../../typography/components/TypographyLegacy/TypographyLegacy';

import stylesCreate from './stylesCreate';

const TooltipTitle: FC<ITitlePopup> = props => {
  const [styles] = useStyles(stylesCreate);
  const {title, titleStyles, titleFont} = props;

  return (
    <TypographyLegacy
      font={titleFont ? titleFont : 'SemiBold-Contrast-XS'}
      style={[styles.title, titleStyles]}>
      {title}
    </TypographyLegacy>
  );
};

export default TooltipTitle;
