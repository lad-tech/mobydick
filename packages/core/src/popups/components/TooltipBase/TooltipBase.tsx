import {FC, useEffect, useMemo, useRef, useState} from 'react';
import {Animated} from 'react-native';
import {useSafeAreaFrame} from 'react-native-safe-area-context';

import {PopupBase} from '../PopupBase';
import {IPosition} from '../../types';
import useStyles from '../../../styles/hooks/useStyles';

import stylesCreate from './stylesCreate';
import TooltipTitle from './TooltipTitle';
import DescriptionText from './DescriptionText';
import Arrow from './Arrow';
import LeftButton from './LeftButton';
import {IPlacement, ITooltipBaseProps} from './types';

const TooltipBase: FC<ITooltipBaseProps> & {
  Title: typeof TooltipTitle;
  DescriptionText: typeof DescriptionText;
  Arrow: typeof Arrow;
  LeftButton: typeof LeftButton;
} = props => {
  const {
    containerStyle,
    children,
    onClose,
    overlayStyle,
    position,
    placement,
    refCurrent,
    timeShow,
  } = props;
  const [styles] = useStyles(stylesCreate);
  const {width, height} = useSafeAreaFrame();

  const [positionValueY, setPositionValueY] = useState(0);
  const [positionValueX, setPositionValueX] = useState(0);

  const timeout = useRef<ReturnType<typeof setTimeout> | null>(null);

  useEffect(() => {
    if (timeShow) {
      timeout.current = setTimeout(() => {
        onClose();
      }, timeShow);
    }

    return () => {
      timeout.current && clearTimeout(timeout.current);
    };
  }, []);

  useMemo(() => {
    refCurrent?.current?.measure((_x, _y, _width, _height, _pageX, pageY) => {
      if (pageY) {
        const androidValue = height - pageY;

        position === IPosition.top
          ? setPositionValueY(pageY + _height)
          : setPositionValueY(androidValue);

        placement === IPlacement.start
          ? setPositionValueX(_pageX)
          : setPositionValueX(width - _pageX - _width);
      }
    });
  }, []);

  if (positionValueY === 0) {
    return null;
  }

  return (
    <PopupBase
      onClose={onClose}
      overlayStyle={[styles.overlayStyle, overlayStyle]}>
      <Animated.View
        style={[
          styles.container,
          containerStyle,
          position === IPosition.top && {
            top: positionValueY,
          },
          position === IPosition.bottom && {
            bottom: positionValueY,
          },
          placement === IPlacement.start && {
            left: positionValueX,
          },
          placement === IPlacement.end && {
            right: positionValueX,
          },
        ]}>
        {children}
      </Animated.View>
    </PopupBase>
  );
};

TooltipBase.Title = TooltipTitle;
TooltipBase.DescriptionText = DescriptionText;
TooltipBase.Arrow = Arrow;
TooltipBase.LeftButton = LeftButton;
export default TooltipBase;
