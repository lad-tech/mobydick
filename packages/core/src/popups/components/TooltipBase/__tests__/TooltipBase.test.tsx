import {render} from '@testing-library/react-native';
import {View} from 'react-native';
import {SafeAreaProvider} from 'react-native-safe-area-context';
import React from 'react';

import TooltipBase from '../TooltipBase';
import {IPlacement} from '../types';
import {IPosition} from '../../../types';
import Button from '../../../../cta/components/Button/Button';
import {ITouchableOpacity} from '../../../../basic/components/TouchableOpacity/types';

describe('@lad-tech/mobydick-core/TooltipBase/TooltipBase', () => {
  jest.useFakeTimers();
  let buttonRef: React.RefObject<ITouchableOpacity>;
  beforeEach(() => {
    buttonRef = React.createRef();
    render(<Button ref={buttonRef} />);
  });
  it('should renders correctly bottom center', () => {
    jest
      .spyOn(buttonRef.current as View, 'measure')
      .mockImplementation(
        (
          cb: (
            x: number,
            y: number,
            width: number,
            height: number,
            pageX: number,
            pageY: number,
          ) => void,
        ) => cb(0, 24, 1, 1, 287, 2410),
      );

    const {toJSON} = render(
      <SafeAreaProvider>
        <TooltipBase
          id={'id'}
          onClose={() => null}
          position={IPosition.bottom}
          placement={IPlacement.center}
          refCurrent={buttonRef}
        />
      </SafeAreaProvider>,
    );

    expect(toJSON()).toMatchSnapshot();
  });
  it('should renders correctly top start', () => {
    const {toJSON} = render(
      <SafeAreaProvider>
        <TooltipBase
          id={'id'}
          onClose={() => null}
          position={IPosition.top}
          placement={IPlacement.start}
          refCurrent={buttonRef}
        />
      </SafeAreaProvider>,
    );

    expect(toJSON()).toMatchSnapshot();
  });
  it('should renders correctly top end', () => {
    const {toJSON} = render(
      <SafeAreaProvider>
        <TooltipBase
          id={'id'}
          onClose={() => null}
          position={IPosition.top}
          placement={IPlacement.end}
          refCurrent={buttonRef}
        />
      </SafeAreaProvider>,
    );

    expect(toJSON()).toMatchSnapshot();
  });
  it('should renders correctly Timeshow', () => {
    const onClose = jest.fn();
    const {toJSON} = render(
      <SafeAreaProvider>
        <TooltipBase
          id={'id'}
          onClose={onClose}
          timeShow={3000}
          position={IPosition.top}
          placement={IPlacement.end}
          refCurrent={buttonRef}
        />
      </SafeAreaProvider>,
    );
    jest.runAllTimers();
    expect(toJSON()).toMatchSnapshot();
  });
});
