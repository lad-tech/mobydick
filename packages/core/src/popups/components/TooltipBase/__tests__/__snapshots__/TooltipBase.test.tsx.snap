// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`@lad-tech/mobydick-core/TooltipBase/TooltipBase should renders correctly Timeshow 1`] = `
<View
  accessibilityState={
    {
      "busy": undefined,
      "checked": undefined,
      "disabled": undefined,
      "expanded": undefined,
      "selected": undefined,
    }
  }
  accessibilityValue={
    {
      "max": undefined,
      "min": undefined,
      "now": undefined,
      "text": undefined,
    }
  }
  accessible={true}
  collapsable={false}
  focusable={true}
  onBlur={[Function]}
  onClick={[Function]}
  onFocus={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    [
      {
        "alignItems": "center",
        "backgroundColor": "rgba(21, 28, 44, 0.2)",
        "height": "100%",
        "position": "absolute",
        "width": "100%",
      },
      [
        {
          "backgroundColor": "transparent",
        },
        undefined,
      ],
    ]
  }
  testID="Pressable"
>
  <View
    collapsable={false}
    style={
      {
        "alignItems": "flex-start",
        "backgroundColor": "#20242D",
        "borderRadius": 8,
        "justifyContent": "center",
        "marginVertical": 8,
        "paddingHorizontal": 16,
        "paddingVertical": 8,
        "position": "absolute",
        "right": 32,
        "top": 2411,
        "zIndex": 1,
      }
    }
  />
</View>
`;

exports[`@lad-tech/mobydick-core/TooltipBase/TooltipBase should renders correctly bottom center 1`] = `
<View
  accessibilityState={
    {
      "busy": undefined,
      "checked": undefined,
      "disabled": undefined,
      "expanded": undefined,
      "selected": undefined,
    }
  }
  accessibilityValue={
    {
      "max": undefined,
      "min": undefined,
      "now": undefined,
      "text": undefined,
    }
  }
  accessible={true}
  collapsable={false}
  focusable={true}
  onBlur={[Function]}
  onClick={[Function]}
  onFocus={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    [
      {
        "alignItems": "center",
        "backgroundColor": "rgba(21, 28, 44, 0.2)",
        "height": "100%",
        "position": "absolute",
        "width": "100%",
      },
      [
        {
          "backgroundColor": "transparent",
        },
        undefined,
      ],
    ]
  }
  testID="Pressable"
>
  <View
    collapsable={false}
    style={
      {
        "alignItems": "flex-start",
        "backgroundColor": "#20242D",
        "borderRadius": 8,
        "bottom": -1770,
        "justifyContent": "center",
        "marginVertical": 8,
        "paddingHorizontal": 16,
        "paddingVertical": 8,
        "position": "absolute",
        "zIndex": 1,
      }
    }
  />
</View>
`;

exports[`@lad-tech/mobydick-core/TooltipBase/TooltipBase should renders correctly top end 1`] = `
<View
  accessibilityState={
    {
      "busy": undefined,
      "checked": undefined,
      "disabled": undefined,
      "expanded": undefined,
      "selected": undefined,
    }
  }
  accessibilityValue={
    {
      "max": undefined,
      "min": undefined,
      "now": undefined,
      "text": undefined,
    }
  }
  accessible={true}
  collapsable={false}
  focusable={true}
  onBlur={[Function]}
  onClick={[Function]}
  onFocus={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    [
      {
        "alignItems": "center",
        "backgroundColor": "rgba(21, 28, 44, 0.2)",
        "height": "100%",
        "position": "absolute",
        "width": "100%",
      },
      [
        {
          "backgroundColor": "transparent",
        },
        undefined,
      ],
    ]
  }
  testID="Pressable"
>
  <View
    collapsable={false}
    style={
      {
        "alignItems": "flex-start",
        "backgroundColor": "#20242D",
        "borderRadius": 8,
        "justifyContent": "center",
        "marginVertical": 8,
        "paddingHorizontal": 16,
        "paddingVertical": 8,
        "position": "absolute",
        "right": 32,
        "top": 2411,
        "zIndex": 1,
      }
    }
  />
</View>
`;

exports[`@lad-tech/mobydick-core/TooltipBase/TooltipBase should renders correctly top start 1`] = `
<View
  accessibilityState={
    {
      "busy": undefined,
      "checked": undefined,
      "disabled": undefined,
      "expanded": undefined,
      "selected": undefined,
    }
  }
  accessibilityValue={
    {
      "max": undefined,
      "min": undefined,
      "now": undefined,
      "text": undefined,
    }
  }
  accessible={true}
  collapsable={false}
  focusable={true}
  onBlur={[Function]}
  onClick={[Function]}
  onFocus={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    [
      {
        "alignItems": "center",
        "backgroundColor": "rgba(21, 28, 44, 0.2)",
        "height": "100%",
        "position": "absolute",
        "width": "100%",
      },
      [
        {
          "backgroundColor": "transparent",
        },
        undefined,
      ],
    ]
  }
  testID="Pressable"
>
  <View
    collapsable={false}
    style={
      {
        "alignItems": "flex-start",
        "backgroundColor": "#20242D",
        "borderRadius": 8,
        "justifyContent": "center",
        "left": 287,
        "marginVertical": 8,
        "paddingHorizontal": 16,
        "paddingVertical": 8,
        "position": "absolute",
        "top": 2411,
        "zIndex": 1,
      }
    }
  />
</View>
`;
